ramips: Add support for Gaoke FG7000N

MightyDuke opened the pull request:
Gaoke FG7000N is a SFP gateway/WiFi router.

Specifications:
CPU: Mediatek MT7620a
Switch: Qualcomm QCA8337
RAM: 128 MiB
Flash: 32 MiB
WiFI: 802.11n, 2x internal antennas
Ports: 4x LAN, 1x WAN, 1x SFP, 1x USB
GPIO: 2x buttons, 15x LEDs, +1 hidden inside,
      4 LAN LEDs, 1 WAN LED and 1 SFP LED are hardwired to the switch
      2 LEDs are labelled "N/A", repurposed as status indicators
UART: 5-pin header, 3.3V 115200 8N1, pins are:
      GND (marked with triangle) VCC TXD RXD GND

Notes:
WiFi is very slow, seems related to:
https://forum.openwrt.org/t/mt7620a-wifi-still-very-slow/145843/6

Back up your firmware! You can download the entire flash over SSH:
1. SSH into the device as "user" (default password "user") over LAN port 1 or WiFi (device is using old SSH, you need to specify -o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedKeyTypes=+ssh-rsa).
2. Type "enable" to enter privileged mode (default password is "super").
3. Start a TFTP server on your PC.
4. Run "tftp <your PC IP> put flash <some filename>".
5. It'll tell you it failed at the end, the file size should be exactly 32 MiB.

Flashing instructions:
1. Connect to UART, plug your PC into a LAN port.
2. Start the device.
3. When asked, press 1 to upload an image to RAM.
4. You will be asked for device IP, server IP and file name.
5. Set a static IP on your PC with the server IP you provided.
6. Start a TFTP server with the "initramfs-kernel" firmware image (use the file name you provided).
7. Wait for it to upload and start OpenWrt.
8. Wait for the "N/A" LED to stop blinking.
9. Switch your PC to DHCP, open 192.168.1.1 in your browser.
10. Flash the "squashfs-sysupgrade" image from there.

neheb commented on target/linux/ramips/dts/mt7620a_gaoke_fg7000n.dts:
```diff
@@ -0,0 +1,219 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0x0>;
```
This should be converted to nvmem.

neheb replied:
https://github.com/openwrt/openwrt/commit/da42c329c6c73b5f4eb40f894195c488ba76d76b for examples

neheb commented on target/linux/ramips/dts/mt7620a_gaoke_fg7000n.dts:
```diff
@@ -0,0 +1,219 @@
+};
+
+&factory {
+	compatible = "nvmem-cells";
```
this should be inlined and use nvmem-layouts.

neheb replied:
I don’t see nvmem-layout

MightyDuke commented:
I made the required changes

MightyDuke commented:
> To improve speed try to add:
> 
> ```
> &wmac {
> 	pinctrl-names = "default", "pa_gpio";
> 	pinctrl-0 = <&pa_pins>;
> 	pinctrl-1 = <&pa_gpio_pins>;
> ```
> 
> Disable WPA3

It seems to have improved a bit in the main branch, I originally tested it on 23.05. WPA3 off, management frame protection off, getting around 20 Mb/s in 20MHz, mode, 40 Mb/s in 40MHz mode.

Also, I see other devices have a `&state_default` section and I'm not sure what it does. Should I add it in?

neheb commented:
Note https://github.com/openwrt/openwrt/pull/16097

MightyDuke committed:
ramips: Add support for Gaoke FG7000N

Gaoke FG7000N is a SFP gateway/WiFi router.

Specifications:
CPU: Mediatek MT7620a
Switch: Qualcomm QCA8337
RAM: 128 MiB
Flash: 32 MiB
WiFI: 802.11n, 2x internal antennas
Ports: 4x LAN, 1x WAN, 1x SFP, 1x USB
GPIO: 2x buttons, 15x LEDs, +1 hidden inside,
      4 LAN LEDs, 1 WAN LED and 1 SFP LED are hardwired to the switch
      2 LEDs are labelled "N/A", repurposed as status indicators
UART: 5-pin header, 3.3V 115200 8N1, pins are:
      GND (marked with triangle) VCC TXD RXD GND

Notes:
WiFi is very slow, seems related to:
https://forum.openwrt.org/t/mt7620a-wifi-still-very-slow/145843/6

Back up your firmware! You can download the entire flash over SSH:
1. SSH into the device as "user" (default password "user") over
   LAN port 1 or WiFi (device is using old SSH, you need to specify
   -o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedKeyTypes=+ssh-rsa).
2. Type "enable" to enter privileged mode (default password is "super").
3. Start a TFTP server on your PC.
4. Run "tftp <your PC IP> put flash <some filename>".
5. It'll tell you it failed at the end, the file size should
   be exactly 32 MiB.

Flashing instructions:
1. Connect to UART, plug your PC into a LAN port.
2. Start the device.
3. When asked, press 1 to upload an image to RAM.
4. You will be asked for device IP, server IP and file name.
5. Set a static IP on your PC with the server IP you provided.
6. Start a TFTP server with the "initramfs-kernel" firmware image
   (use the file name you provided).
7. Wait for it to upload and start OpenWrt.
8. Wait for the "N/A" LED to stop blinking.
9. Switch your PC to DHCP, open 192.168.1.1 in your browser.
10. Flash the "squashfs-sysupgrade" image from there.

Signed-off-by: Josef Macháč <papa.the.roach@gmail.com>

namiltd commented:
Code like:
```
&state_default {
	gpio {
		groups = "i2c", "uartf", "ephy", "rgmii2";
		function = "gpio";
	};
};
```
define which outputs are to be switched to GPIO type outputs
